api/src/middleware/error-handler.ts

```diff
@@ -1,105 +1,118 @@
-import { ErrorCode, MethodNotAllowedError, isDirectusError } from '@directus/errors';
-import { isObject, toArray } from '@directus/utils';
+import { ErrorCode, InternalServerError, isDirectusError } from '@directus/errors';
+import type { DeepPartial } from '@directus/types';
+import { isObject } from '@directus/utils';
 import { getNodeEnv } from '@directus/utils/node';
-import type { ErrorRequestHandler } from 'express';
+import type { ErrorRequestHandler, NextFunction, Request, Response } from 'express';
 import getDatabase from '../database/index.js';
 import emitter from '../emitter.js';
 import { useLogger } from '../logger.js';
 
-// Note: keep all 4 parameters here. That's how Express recognizes it's the error handler, even if
-// we don't use next
-const errorHandler: ErrorRequestHandler = (err, req, res, _next) => {
-	const logger = useLogger();
-
-	let payload: any = {
-		errors: [],
+type ApiError = {
+	message: string;
+	extensions: {
+		code: string;
+		[key: string]: any;
 	};
+};
 
-	const errors = toArray<unknown>(err);
+const FALLBACK_ERROR = new InternalServerError();
+
+export const errorHandler = asyncErrorHandler(async (err, req, res) => {
+	const logger = useLogger();
 
+	let errors: ApiError[] = [];
 	let status: number | null = null;
 
-	for (const error of errors) {
-		if (getNodeEnv() === 'development') {
-			if (isObject(error)) {
-				error['extensions'] = {
-					...(error['extensions'] || {}),
-					stack: error['stack'],
-				};
-			}
+	// It can be assumed that at least one error is given
+	const receivedErrors: unknown[] = Array.isArray(err) ? err : [err];
+
+	for (const error of receivedErrors) {
+		// In dev mode, if available, expose stack trace under error's extensions data
+		if (getNodeEnv() === 'development' && error instanceof Error && error.stack) {
+			((error as DeepPartial<ApiError>).extensions ??= {})['stack'] = error.stack;
 		}
 
 		if (isDirectusError(error)) {
 			logger.debug(error);
 
-			if (!status) {
+			if (status === null) {
+				// Use current error status as response status
 				status = error.status;
 			} else if (status !== error.status) {
-				status = 500;
+				// Fallback if status has already been set by a preceding error
+				// and doesn't match the current one
+				status = FALLBACK_ERROR.status;
 			}
 
-			payload.errors.push({
+			errors.push({
 				message: error.message,
 				extensions: {
+					// Expose error code under error's extensions data
 					code: error.code,
 					...(error.extensions ?? {}),
```

Comment:
```suggestion
				extensions: {
					...(error.extensions ?? {}),
					code: error.code,
```
Inverted the order to ensure the `error.extensions` spreading cannot overwrite the set code.